crates/ruma-events/src/rtc/notification.rs

```diff
@@ -0,0 +1,204 @@
+//! Unstable: `org.matrix.msc4075.rtc.notification`
+
+use js_int::UInt;
+use ruma_common::MilliSecondsSinceUnixEpoch;
+use ruma_events::{relation::Reference, Mentions};
+use ruma_macros::EventContent;
+use serde::{Deserialize, Serialize};
+
+/// The content of an `m.rtc.notification` event.
+#[derive(Clone, Debug, Deserialize, Serialize, EventContent)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+#[ruma_event(
+    type = "m.rtc.notification",
+    kind = MessageLike
+)]
+pub struct RtcNotificationEventContent {
+    /// Local timestamp observed by the sender device.
+    ///
+    /// Used with `lifetime` to determine validity; receivers SHOULD compare with
+    /// `origin_server_ts` and prefer it if the difference is large.
+    pub sender_ts: MilliSecondsSinceUnixEpoch,
+
+    /// Relative time from `sender_ts` during which the notification is considered valid.
+    pub lifetime: UInt,
+
+    /// Intentional mentions determining who should be notified.
+    #[serde(rename = "m.mentions", default, skip_serializing_if = "Option::is_none")]
+    pub mentions: Option<Mentions>,
+
+    /// Optional reference to the related `m.rtc.member` event.
+    #[serde(rename = "m.relates_to", skip_serializing_if = "Option::is_none")]
+    pub relates_to: Option<Reference>,
+
+    /// How this notification should notify the receiver.
+    #[serde(rename = "notification_type")]
+    pub notification_type: NotificationType,
+}
+
+impl RtcNotificationEventContent {
+    /// Creates a new `RtcNotificationEventContent` with the given configuration.
+    pub fn new(
+        sender_ts: MilliSecondsSinceUnixEpoch,
+        lifetime: UInt,
+        notification_type: NotificationType,
+    ) -> Self {
+        Self { sender_ts, lifetime, mentions: None, relates_to: None, notification_type }
+    }
+
+    /// Calculates the timestamp at which this notification is considered invalid.
+    /// This calculation is based on MSC4075 and tries to use the `sender_ts` as the starting point
+    /// and the `lifetime` as the duration for which the notification is valid.
+    ///
+    /// The `sender_ts` cannot be trusted since it is a generated value by the sending client.
+    /// To mitigate issue because of misconfigured client clocks, the MSC requires
+    /// that the `origin_server_ts` is used as the starting point if the difference is large.
+    ///
+    /// # Arguments:
+    ///
+    /// - `max_sender_ts_offset` is the maximum allowed offset between the two timestamps. (default
+    /// 20s)
+    /// - `origin_server_ts` has to be set to the origin_server_ts from the event containing this
+    ///   event content.
+    ///
+    /// # Examples
+    /// To start a timer until this client should stop ringing for this notification:
+    /// ```
+    /// let duration_ring = MilliSecondsSinceUnixEpoch::now()
+    ///     .saturated_sub(content.invalid_ts(event.origin_server_ts(), None));
+    /// ```
+    pub fn invalid_ts(
```

Comment:
Wouldn't "expired" be a clearer name?

Also is the timestamp something that users need, or do they actually just want to now whether the event is expired?

Reply:
They need to start a timer to know when they can stop the ringing.

Reply:
I was considering a method that computes a relative expiration duration but it felt more affirming to have a method returning the ts and compute the duration just before the timer is started.